<a href="https://colab.research.google.com/github/HamburgermanA3445/heartdiseaseprediction/blob/main/HeartDisease_%5B5.Best%5D.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

Unsuccessful, obtained from Kaggle:

https://www.kaggle.com/datasets/alexteboul/heart-disease-health-indicators-dataset

*some of the cells at the end were not run

In [ ]:
!unzip "/content/alex_teboul (5).zip"

Archive:  /content/alex_teboul (5).zip
  inflating: 2020/heart_2020_cleaned.csv  
  inflating: 2022/heart_2022_no_nans.csv  
  inflating: 2022/heart_2022_with_nans.csv  


In [ ]:
import pandas as pd
from sklearn.preprocessing import LabelEncoder

In [ ]:
heart_data = pd.read_csv("/content/2022/heart_2022_no_nans.csv")

In [ ]:
label_encoders = {}

for col in heart_data.select_dtypes(include=['object']).columns:
    le = LabelEncoder()
    heart_data[col] = le.fit_transform(heart_data[col])
    label_encoders[col] = le

In [ ]:
heart_data.columns

Index(['State', 'Sex', 'GeneralHealth', 'PhysicalHealthDays',
       'MentalHealthDays', 'LastCheckupTime', 'PhysicalActivities',
       'SleepHours', 'RemovedTeeth', 'HadHeartAttack', 'HadAngina',
       'HadStroke', 'HadAsthma', 'HadSkinCancer', 'HadCOPD',
       'HadDepressiveDisorder', 'HadKidneyDisease', 'HadArthritis',
       'HadDiabetes', 'DeafOrHardOfHearing', 'BlindOrVisionDifficulty',
       'DifficultyConcentrating', 'DifficultyWalking',
       'DifficultyDressingBathing', 'DifficultyErrands', 'SmokerStatus',
       'ECigaretteUsage', 'ChestScan', 'RaceEthnicityCategory', 'AgeCategory',
       'HeightInMeters', 'WeightInKilograms', 'BMI', 'AlcoholDrinkers',
       'HIVTesting', 'FluVaxLast12', 'PneumoVaxEver', 'TetanusLast10Tdap',
       'HighRiskLastYear', 'CovidPos'],
      dtype='object')

In [ ]:
target = 'HadHeartAttack'
all_columns = heart_data.columns.drop(target)

In [ ]:
heart_data.head()

State  Sex  GeneralHealth  PhysicalHealthDays  MentalHealthDays  \
0      0    0              4                 4.0               0.0   
1      0    1              4                 0.0               0.0   
2      0    1              4                 0.0               0.0   
3      0    0              1                 5.0               0.0   
4      0    0              2                 3.0              15.0   

   LastCheckupTime  PhysicalActivities  SleepHours  RemovedTeeth  \
0                3                   1         9.0             3   
1                3                   1         6.0             3   
2                3                   0         8.0             1   
3                3                   1         9.0             3   
4                3                   1         5.0             0   

   HadHeartAttack  ...  HeightInMeters  WeightInKilograms    BMI  \
0               0  ...            1.60              71.67  27.99   
1               0  ...            1.78              95.25  30.13   
2               0  ...            1.85             108.86  31.66   
3               0  ...            1.70              90.72  31.32   
4               0  ...            1.55              79.38  33.07   

   AlcoholDrinkers  HIVTesting  FluVaxLast12  PneumoVaxEver  \
0                0           0             1              1   
1                0           0             1              1   
2                1           0             0              1   
3                0           0             1              1   
4                0           0             1              1   

   TetanusLast10Tdap  HighRiskLastYear  CovidPos  
0                  1                 0         0  
1                  2                 0         0  
2                  0                 0         2  
3                  0                 0         2  
4                  0                 0         0  

[5 rows x 40 columns]

#Logistic Model

In [ ]:
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression

In [ ]:
#Use for filtering categorical data into numerical
for l in all_columns:
  all_x = []
  for x in heart_data[l]:
    if x not in all_x:
      all_x.append(x)
  print(f'{l}: {all_x}')

State: [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35, 36, 37, 38, 39, 41, 42, 43, 44, 45, 46, 47, 49, 50, 51, 52, 53, 11, 40, 48]
Sex: [0, 1]
GeneralHealth: [4, 1, 2, 0, 3]
PhysicalHealthDays: [4.0, 0.0, 5.0, 3.0, 2.0, 25.0, 30.0, 15.0, 29.0, 8.0, 16.0, 20.0, 10.0, 9.0, 7.0, 1.0, 21.0, 6.0, 27.0, 14.0, 12.0, 11.0, 13.0, 28.0, 17.0, 23.0, 24.0, 26.0, 18.0, 22.0, 19.0]
MentalHealthDays: [0.0, 15.0, 4.0, 25.0, 5.0, 30.0, 27.0, 3.0, 2.0, 1.0, 10.0, 20.0, 21.0, 6.0, 7.0, 8.0, 14.0, 9.0, 12.0, 18.0, 29.0, 28.0, 17.0, 11.0, 16.0, 13.0, 26.0, 22.0, 24.0, 19.0, 23.0]
LastCheckupTime: [3, 0, 1, 2]
PhysicalActivities: [1, 0]
SleepHours: [9.0, 6.0, 8.0, 5.0, 7.0, 10.0, 4.0, 12.0, 3.0, 18.0, 11.0, 2.0, 1.0, 16.0, 14.0, 15.0, 13.0, 20.0, 24.0, 23.0, 19.0, 17.0, 22.0]
RemovedTeeth: [3, 1, 0, 2]
HadAngina: [0, 1]
HadStroke: [0, 1]
HadAsthma: [0, 1]
HadSkinCancer: [0, 1]
HadCOPD: [0, 1]
HadDepressiveDisorder: [0, 1]
Had

In [ ]:
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt
from sklearn.metrics import accuracy_score, f1_score

In [ ]:
X = heart_data[all_columns]
y = heart_data[target]
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)
lg = LogisticRegression()
lg.fit(X_train, y_train)
y_pred = lg.predict(X_test)
accuracy_score(y_test, y_pred)

/usr/local/lib/python3.10/dist-packages/sklearn/linear_model/_logistic.py:458: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


0.948135352098364

In [ ]:
from sklearn.metrics import classification_report

print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       0.96      0.99      0.97     46573
           1       0.53      0.24      0.33      2632

    accuracy                           0.95     49205
   macro avg       0.75      0.61      0.65     49205
weighted avg       0.94      0.95      0.94     49205



In [ ]:
coefficients = lg.coef_[0]
feature_importance = np.abs(coefficients)

importance = pd.DataFrame({
    'Feature': X.columns,
    'Coefficient': coefficients,
    'Importance': feature_importance
})

importance = importance.sort_values(by='Importance', ascending=False)

In [ ]:
importance

Feature  Coefficient  Importance
9                   HadAngina     2.486774    2.486774
26                  ChestScan     0.781615    0.781615
29             HeightInMeters    -0.780024    0.780024
10                  HadStroke     0.675792    0.675792
1                         Sex     0.508067    0.508067
32            AlcoholDrinkers    -0.395502    0.395502
6          PhysicalActivities    -0.383399    0.383399
21          DifficultyWalking     0.273084    0.273084
35              PneumoVaxEver     0.245766    0.245766
15           HadKidneyDisease     0.241308    0.241308
13                    HadCOPD     0.234399    0.234399
17                HadDiabetes     0.223421    0.223421
25            ECigaretteUsage    -0.219622    0.219622
24               SmokerStatus    -0.195155    0.195155
18        DeafOrHardOfHearing     0.194915    0.194915
16               HadArthritis     0.174479    0.174479
5             LastCheckupTime    -0.133747    0.133747
19    BlindOrVisionDifficulty     0.132257    0.132257
33                 HIVTesting    -0.127803    0.127803
23          DifficultyErrands     0.096876    0.096876
8                RemovedTeeth    -0.095259    0.095259
28                AgeCategory     0.081929    0.081929
7                  SleepHours    -0.074921    0.074921
2               GeneralHealth    -0.074562    0.074562
34               FluVaxLast12    -0.073783    0.073783
27      RaceEthnicityCategory    -0.072525    0.072525
20    DifficultyConcentrating     0.072025    0.072025
31                        BMI    -0.063750    0.063750
22  DifficultyDressingBathing     0.053187    0.053187
11                  HadAsthma     0.049484    0.049484
37           HighRiskLastYear    -0.045528    0.045528
36          TetanusLast10Tdap    -0.035559    0.035559
12              HadSkinCancer     0.035201    0.035201
14      HadDepressiveDisorder     0.021109    0.021109
38                   CovidPos    -0.013437    0.013437
30          WeightInKilograms     0.011394    0.011394
0                       State    -0.003121    0.003121
3          PhysicalHealthDays    -0.002442    0.002442
4            MentalHealthDays     0.002055    0.002055

In [ ]:
a_dict = {}
for p in all_columns:
  X = heart_data[[p]]
  y = heart_data[target]
  X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)
  lg = LogisticRegression()
  lg.fit(X_train, y_train)
  y_pred = lg.predict(X_test)
  accuracy = accuracy_score(y_test, y_pred)
  f1 = f1_score(y_test, y_pred)
  a_dict[p] = accuracy

a_dict = sorted(a_dict.items(), key= lambda x: x[1])
print(f'{a_dict[-1][0]} had the highest score of {a_dict[-1][1]}')

CovidPos had the highest score of 0.9465095010669647


In [ ]:
sorted(a_dict, key=lambda x: x[1], reverse=True)

[('State', 0.9465095010669647),
 ('Sex', 0.9465095010669647),
 ('GeneralHealth', 0.9465095010669647),
 ('PhysicalHealthDays', 0.9465095010669647),
 ('MentalHealthDays', 0.9465095010669647),
 ('LastCheckupTime', 0.9465095010669647),
 ('PhysicalActivities', 0.9465095010669647),
 ('SleepHours', 0.9465095010669647),
 ('RemovedTeeth', 0.9465095010669647),
 ('HadAngina', 0.9465095010669647),
 ('HadStroke', 0.9465095010669647),
 ('HadAsthma', 0.9465095010669647),
 ('HadSkinCancer', 0.9465095010669647),
 ('HadCOPD', 0.9465095010669647),
 ('HadDepressiveDisorder', 0.9465095010669647),
 ('HadKidneyDisease', 0.9465095010669647),
 ('HadArthritis', 0.9465095010669647),
 ('HadDiabetes', 0.9465095010669647),
 ('DeafOrHardOfHearing', 0.9465095010669647),
 ('BlindOrVisionDifficulty', 0.9465095010669647),
 ('DifficultyConcentrating', 0.9465095010669647),
 ('DifficultyWalking', 0.9465095010669647),
 ('DifficultyDressingBathing', 0.9465095010669647),
 ('DifficultyErrands', 0.9465095010669647),
 ('SmokerSt

In [ ]:
#Testing a combination of the most important features ranked by the coefficients
X = heart_data[['exercise angina', 'sex', 'ST slope', 'fasting blood sugar', 'chest pain type', 'oldpeak']]
y = heart_data[target]

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

lg = LogisticRegression()
lg.fit(X_train, y_train)
y_pred = lg.predict(X_test)

KeyError: "None of [Index(['exercise angina', 'sex', 'ST slope', 'fasting blood sugar',\n       'chest pain type', 'oldpeak'],\n      dtype='object')] are in the [columns]"

In [ ]:
print(classification_report(y_test, y_pred))

In [ ]:
print(heart_data[target].value_counts())

target
1    629
0    561
Name: count, dtype: int64


#Neural Network

In [ ]:
import numpy as np
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score, confusion_matrix, f1_score
from sklearn.preprocessing import StandardScaler

In [ ]:
all_columns = heart_data.columns.drop(target)

In [ ]:
X = heart_data[all_columns].values
y = heart_data[target].values

In [ ]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

In [ ]:
scaler = StandardScaler()
X_train = scaler.fit_transform(X_train)
X_test = scaler.transform(X_test)

In [ ]:
class NeuralNetwork:
    def __init__(self, num_feats):
        np.random.seed(1)
        self.synaptic_weights = 2 * np.random.random((num_feats, 1)) - 1

    def sigmoid(self, x):
        return 1 / (1 + np.exp(-x))

    def sigmoid_derivative(self, x):
        return x * (1 - x)

    def train(self, training_inputs, training_outputs, training_iterations):
        for iteration in range(training_iterations):
            output = self.think(training_inputs)
            error = training_outputs - output
            adjustments = np.dot(training_inputs.T, error * self.sigmoid_derivative(output))
            self.synaptic_weights += adjustments

    def think(self, inputs):
        inputs = inputs.astype(float)
        output = self.sigmoid(np.dot(inputs, self.synaptic_weights))
        return output

In [ ]:
neural_network = NeuralNetwork(num_feats=len(all_columns))

print(f"Random starting synaptic weights: \n{neural_network.synaptic_weights}")

training_inputs = X_train
training_outputs = y_train.reshape(-1, 1)

neural_network.train(training_inputs, training_outputs, 1000)

print("Synaptic weights after training: ")
print(neural_network.synaptic_weights)

Random starting synaptic weights: 
[[-0.16595599]
 [ 0.44064899]
 [-0.99977125]
 [-0.39533485]
 [-0.70648822]
 [-0.81532281]
 [-0.62747958]
 [-0.30887855]
 [-0.20646505]
 [ 0.07763347]
 [-0.16161097]]
Synaptic weights after training: 
[[ 24.846768  ]
 [ 31.74681265]
 [ 42.58894145]
 [  7.06035297]
 [-17.84505095]
 [ 15.92746448]
 [  6.41169229]
 [-36.45599246]
 [ 42.53868126]
 [ 43.75605679]
 [ 47.40503255]]


In [ ]:
test_inputs = X_test
test_outputs = y_test.reshape(-1, 1)

raw_test_pred = neural_network.think(test_inputs)
test_pred = np.round(raw_test_pred).astype(int)
accuracy = accuracy_score(test_outputs, test_pred)

print(f'Accuracy: {accuracy_score(test_outputs, test_pred)}')
print(classification_report(y_test, test_pred))

Accuracy: 0.8445378151260504
              precision    recall  f1-score   support

           0       0.83      0.82      0.83       107
           1       0.86      0.86      0.86       131

    accuracy                           0.84       238
   macro avg       0.84      0.84      0.84       238
weighted avg       0.84      0.84      0.84       238



#Tensorflow

In [ ]:
import tensorflow as tf
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense

In [ ]:
heart_data.columns

Index(['age', 'sex', 'chest pain type', 'resting bp s', 'cholesterol',
       'fasting blood sugar', 'resting ecg', 'max heart rate',
       'exercise angina', 'oldpeak', 'ST slope', 'target'],
      dtype='object')

In [ ]:
tf.random.set_seed(42)

In [ ]:
def remov(*items):
    return [item for item in list(heart_data.columns.drop(target)) if item not in items]

In [ ]:
all_columns = remov()

In [ ]:
X = heart_data[all_columns]
y = heart_data[target]

In [ ]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

In [ ]:
ts = tf.keras.Sequential([
    tf.keras.layers.Dense(64, activation='relu', input_shape=(len(all_columns),)),
    tf.keras.layers.Dense(32, activation='relu'),
    tf.keras.layers.Dense(1, activation='sigmoid')
])

ts.compile(optimizer=tf.keras.optimizers.Adam(learning_rate=0.001), loss='binary_crossentropy', metrics=['accuracy'])
ts.fit(X_train, y_train, epochs=100, batch_size=10)

Epoch 1/100
96/96 [==============================] - 1s 2ms/step - loss: 0.9693 - accuracy: 0.6439
Epoch 2/100
96/96 [==============================] - 0s 2ms/step - loss: 0.6179 - accuracy: 0.7048
Epoch 3/100
96/96 [==============================] - 0s 2ms/step - loss: 0.5691 - accuracy: 0.7080
Epoch 4/100
96/96 [==============================] - 0s 2ms/step - loss: 0.6885 - accuracy: 0.6828
Epoch 5/100
96/96 [==============================] - 0s 2ms/step - loss: 0.5501 - accuracy: 0.7500
Epoch 6/100
96/96 [==============================] - 0s 2ms/step - loss: 0.5391 - accuracy: 0.7626
Epoch 7/100
96/96 [==============================] - 0s 3ms/step - loss: 0.5831 - accuracy: 0.7416
Epoch 8/100
96/96 [==============================] - 0s 4ms/step - loss: 0.5447 - accuracy: 0.7721
Epoch 9/100
96/96 [==============================] - 0s 4ms/step - loss: 0.6686 - accuracy: 0.7416
Epoch 10/100
96/96 [==============================] - 0s 4ms/step - loss: 0.5336 - accuracy: 0.7689
Epoch 11/

In [ ]:
loss, accuracy = ts.evaluate(X_test, y_test)
print(f"Accuracy: {accuracy*100:.2f}%")

8/8 [==============================] - 0s 4ms/step - loss: 0.4375 - accuracy: 0.8151
Accuracy: 81.51%


In [ ]:
from sklearn.metrics import classification_report

y_pred = (ts.predict(X_test) > 0.5).astype(int)
print(classification_report(y_test, y_pred))

Exception ignored in: <function _xla_gc_callback at 0x7fcc89f28790>
Traceback (most recent call last):
  File "/usr/local/lib/python3.10/dist-packages/jax/_src/lib/__init__.py", line 98, in _xla_gc_callback
    def _xla_gc_callback(*args):
KeyboardInterrupt: 


8/8 [==============================] - 0s 2ms/step
              precision    recall  f1-score   support

           0       0.81      0.78      0.79       107
           1       0.82      0.85      0.83       131

    accuracy                           0.82       238
   macro avg       0.81      0.81      0.81       238
weighted avg       0.81      0.82      0.81       238



With all features: 82%

Without the few least important features evaluated from logistic model: 83%

#Other Models

In [ ]:
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt
from sklearn.model_selection import train_test_split
from sklearn.metrics import classification_report, confusion_matrix, f1_score, accuracy_score, mean_squared_error, r2_score, recall_score, precision_score
from sklearn.linear_model import LogisticRegression
from sklearn.ensemble import RandomForestClassifier
from sklearn.ensemble import ExtraTreesClassifier
from sklearn.tree import DecisionTreeClassifier
from sklearn.neighbors import KNeighborsClassifier
from sklearn.svm import SVC
from sklearn.ensemble import GradientBoostingClassifier
from sklearn.model_selection import GridSearchCV

In [ ]:
heart_data.columns.drop(target)

Index(['age', 'sex', 'chest pain type', 'resting bp s', 'cholesterol',
       'fasting blood sugar', 'resting ecg', 'max heart rate',
       'exercise angina', 'oldpeak', 'ST slope'],
      dtype='object')

In [ ]:
all_columns = list(heart_data.columns.drop(target))
X = heart_data[all_columns]
y = heart_data[target]

In [ ]:
#initializing 5 models
rf = RandomForestClassifier(max_depth=4,random_state=42, class_weight='balanced')
et = ExtraTreesClassifier(n_estimators=100, random_state=42, class_weight='balanced')
dt = DecisionTreeClassifier(max_depth=5, random_state=42, class_weight='balanced')
knn = KNeighborsClassifier(n_neighbors=5)
svm = SVC(kernel='linear', C=1, random_state=42)
gbm = GradientBoostingClassifier(n_estimators=100, learning_rate=0.1, random_state=42)

In [ ]:
#training and testing results lists to store in
accuracy = []
f1 = []
precision = []
recall = []
#splitting data
aX_train, aX_test, ay_train, ay_test = train_test_split(X, y, test_size=0.2, random_state=42)

In [ ]:
#function for training and testing
def training(model, aX_train, aX_test, ay_train, ay_test):
    model.fit(aX_train, ay_train)
    ay_train_pred = model.predict(aX_train)
    ay_test_pred = model.predict(aX_test)
    accuracy.append(accuracy_score(ay_test, ay_test_pred))
    f1.append(f1_score(ay_test, ay_test_pred))
    precision.append(precision_score(ay_test, ay_test_pred))
    recall.append(recall_score(ay_test, ay_test_pred))
    print(classification_report(ay_test, ay_test_pred))

In [ ]:
#calling training function for each model
training(rf, aX_train, aX_test, ay_train, ay_test)

              precision    recall  f1-score   support

           0       0.87      0.87      0.87       107
           1       0.89      0.89      0.89       131

    accuracy                           0.88       238
   macro avg       0.88      0.88      0.88       238
weighted avg       0.88      0.88      0.88       238



In [ ]:
training(dt, aX_train, aX_test, ay_train, ay_test)

              precision    recall  f1-score   support

           0       0.90      0.75      0.82       107
           1       0.82      0.93      0.87       131

    accuracy                           0.85       238
   macro avg       0.86      0.84      0.84       238
weighted avg       0.85      0.85      0.85       238



In [ ]:
training(knn, aX_train, aX_test, ay_train, ay_test)

              precision    recall  f1-score   support

           0       0.68      0.71      0.69       107
           1       0.75      0.73      0.74       131

    accuracy                           0.72       238
   macro avg       0.72      0.72      0.72       238
weighted avg       0.72      0.72      0.72       238



In [ ]:
training(gbm, aX_train, aX_test, ay_train, ay_test)

              precision    recall  f1-score   support

           0       0.91      0.90      0.91       107
           1       0.92      0.93      0.92       131

    accuracy                           0.92       238
   macro avg       0.92      0.91      0.91       238
weighted avg       0.92      0.92      0.92       238



In [ ]:
training(svm, aX_train, aX_test, ay_train, ay_test)

              precision    recall  f1-score   support

           0       0.84      0.82      0.83       107
           1       0.86      0.87      0.86       131

    accuracy                           0.85       238
   macro avg       0.85      0.85      0.85       238
weighted avg       0.85      0.85      0.85       238



In [ ]:
training(et, aX_train, aX_test, ay_train, ay_test)

              precision    recall  f1-score   support

           0       0.92      0.94      0.93       107
           1       0.95      0.93      0.94       131

    accuracy                           0.94       238
   macro avg       0.94      0.94      0.94       238
weighted avg       0.94      0.94      0.94       238



In [ ]:
#all results
models = ["Random Forest", "Decision Tree", "k-Nearest Neighbors", "Gradient Boosting Machine", "Support Vector Machine", "Extra Trees"]
results = pd.DataFrame({
    "Model": models,
    "Accuracy": accuracy,
    "Precision": precision,
    "Recall": recall,
    "F1": f1
})
print(results)

                       Model  Accuracy  Precision    Recall        F1
0              Random Forest  0.882353   0.893130  0.893130  0.893130
1              Decision Tree  0.848739   0.818792  0.931298  0.871429
2        k-Nearest Neighbors  0.718487   0.753968  0.725191  0.739300
3  Gradient Boosting Machine  0.915966   0.917293  0.931298  0.924242
4     Support Vector Machine  0.848739   0.857143  0.870229  0.863636
5                Extra Trees  0.936975   0.953125  0.931298  0.942085


In [ ]:
heart_data.columns

Index(['age', 'sex', 'chest pain type', 'resting bp s', 'cholesterol',
       'fasting blood sugar', 'resting ecg', 'max heart rate',
       'exercise angina', 'oldpeak', 'ST slope', 'target'],
      dtype='object')

##Feature Importance Ranking

In [ ]:
def rank(the_model):
  feature_importances = the_model.feature_importances_
  ranking = pd.DataFrame({
      'Feature': X.columns,
      'Importance': feature_importances
  }).sort_values(by='Importance', ascending=False)

  print(ranking)

In [ ]:
rank(rf)

                Feature  Importance
10             ST slope    0.307498
2       chest pain type    0.192244
9               oldpeak    0.122014
8       exercise angina    0.117147
7        max heart rate    0.067680
4           cholesterol    0.064525
1                   sex    0.042679
0                   age    0.042556
3          resting bp s    0.023041
5   fasting blood sugar    0.013083
6           resting ecg    0.007532


In [ ]:
rank(dt)

                Feature  Importance
10             ST slope    0.513429
2       chest pain type    0.172394
9               oldpeak    0.084700
7        max heart rate    0.083859
1                   sex    0.057109
3          resting bp s    0.029284
5   fasting blood sugar    0.020572
0                   age    0.020384
8       exercise angina    0.009420
4           cholesterol    0.008849
6           resting ecg    0.000000


In [ ]:
rank(gbm)

                Feature  Importance
10             ST slope    0.399717
2       chest pain type    0.141090
9               oldpeak    0.084093
7        max heart rate    0.083209
4           cholesterol    0.077786
8       exercise angina    0.047677
0                   age    0.046599
1                   sex    0.046086
3          resting bp s    0.038352
5   fasting blood sugar    0.023777
6           resting ecg    0.011616


In [ ]:
rank(et)

                Feature  Importance
10             ST slope    0.161499
8       exercise angina    0.140620
2       chest pain type    0.129899
9               oldpeak    0.095566
7        max heart rate    0.093239
4           cholesterol    0.089776
0                   age    0.080637
3          resting bp s    0.075222
1                   sex    0.055412
6           resting ecg    0.041936
5   fasting blood sugar    0.036194


In [ ]:
from sklearn.inspection import permutation_importance

In [ ]:
knn_result = permutation_importance(knn, X_test, y_test, n_repeats=10, random_state=42)

In [ ]:
svm_result = permutation_importance(svm, X_test, y_test, n_repeats=10, random_state=42)

In [ ]:
def rank_2(result):
  perms = pd.DataFrame({
      'Feature': X.columns,
      '(Mean) Importance': result.importances_mean,
  })
  sorted_importances = perms.sort_values(by='(Mean) Importance', ascending=False)

  print(sorted_importances)

In [ ]:
rank_2(knn_result)

                Feature  (Mean) Importance
7        max heart rate           0.137815
4           cholesterol           0.101261
3          resting bp s           0.042437
0                   age           0.020588
8       exercise angina           0.003361
6           resting ecg           0.002941
9               oldpeak           0.002941
1                   sex           0.001681
10             ST slope           0.000000
5   fasting blood sugar          -0.000840
2       chest pain type          -0.002941


In [ ]:
rank_2(svm_result)

                Feature  (Mean) Importance
10             ST slope           0.079412
2       chest pain type           0.037815
8       exercise angina           0.035294
1                   sex           0.025210
9               oldpeak           0.015966
4           cholesterol           0.003361
5   fasting blood sugar           0.001261
6           resting ecg          -0.002521
3          resting bp s          -0.002941
7        max heart rate          -0.005042
0                   age          -0.013866


##Testing collective performance by removing certain features from all features

In [ ]:
initial_dict = {} #run this only once

In [ ]:
def remov(*items):
    return [item for item in list(heart_data.columns.drop(target)) if item not in items]

In [ ]:
all_columns = remov('exercise angina') #'slope' had a good effect on GBM

In [ ]:
X = heart_data[all_columns]
y = heart_data[target]
gbm = GradientBoostingClassifier(n_estimators=100, learning_rate=0.1, random_state=42)
#training and testing results lists to store in
accuracy = []
f1 = []
precision = []
recall = []
#splitting data
aX_train, aX_test, ay_train, ay_test = train_test_split(X, y, test_size=0.2, random_state=42)
#function for training and testing
def training(model, aX_train, aX_test, ay_train, ay_test):
    model.fit(aX_train, ay_train)
    ay_train_pred = model.predict(aX_train)
    ay_test_pred = model.predict(aX_test)
    accuracy.append(accuracy_score(ay_test, ay_test_pred))
    f1.append(f1_score(ay_test, ay_test_pred))
    precision.append(precision_score(ay_test, ay_test_pred))
    recall.append(recall_score(ay_test, ay_test_pred))
training(rf, aX_train, aX_test, ay_train, ay_test)
training(dt, aX_train, aX_test, ay_train, ay_test)
training(knn, aX_train, aX_test, ay_train, ay_test)
training(gbm, aX_train, aX_test, ay_train, ay_test)
training(svm, aX_train, aX_test, ay_train, ay_test)
training(et, aX_train, aX_test, ay_train, ay_test)
#one model's results over different features
models = ["Random Forest", "Decision Tree", "k-Nearest Neighbors", "Gradient Boosting Machine", "Support Vector Machine", "Extra Trees"]
results = pd.DataFrame({
    "Model": models,
    "Accuracy": accuracy,
    "Precision": precision,
    "Recall": recall,
    "F1": f1
})
print(results)
initial_dict[str(all_columns)+" | length "+str(len(all_columns))] = results

                       Model  Accuracy  Precision    Recall        F1
0              Random Forest  0.886555   0.893939  0.900763  0.897338
1              Decision Tree  0.848739   0.818792  0.931298  0.871429
2        k-Nearest Neighbors  0.718487   0.753968  0.725191  0.739300
3  Gradient Boosting Machine  0.907563   0.903704  0.931298  0.917293
4     Support Vector Machine  0.831933   0.837037  0.862595  0.849624
5                Extra Trees  0.936975   0.946154  0.938931  0.942529


In [ ]:
initial_dict

{"['age', 'sex', 'chest pain type', 'resting bp s', 'cholesterol', 'fasting blood sugar', 'resting ecg', 'max heart rate', 'exercise angina', 'oldpeak', 'ST slope']":                        Model  Accuracy  Precision    Recall        F1
 0              Random Forest  0.882353   0.893130  0.893130  0.893130
 1              Decision Tree  0.848739   0.818792  0.931298  0.871429
 2        k-Nearest Neighbors  0.718487   0.753968  0.725191  0.739300
 3  Gradient Boosting Machine  0.915966   0.917293  0.931298  0.924242
 4     Support Vector Machine  0.848739   0.857143  0.870229  0.863636
 5                Extra Trees  0.936975   0.953125  0.931298  0.942085,
 "['age', 'chest pain type', 'resting bp s', 'cholesterol', 'resting ecg', 'max heart rate', 'exercise angina', 'oldpeak', 'ST slope']":                        Model  Accuracy  Precision    Recall        F1
 0              Random Forest  0.865546   0.866667  0.893130  0.879699
 1              Decision Tree  0.815126   0.791946  0.9007

Looks like removing exercise angina increased the accuracy by a tiny bit

##Testing for best single feature for the best models



In [ ]:
#initializing 5 models
rf = RandomForestClassifier(max_depth=4,random_state=42, class_weight='balanced')
et = ExtraTreesClassifier(n_estimators=100, random_state=42, class_weight='balanced')
dt = DecisionTreeClassifier(max_depth=5, random_state=42, class_weight='balanced')
knn = KNeighborsClassifier(n_neighbors=5)
svm = SVC(kernel='linear', C=1, random_state=42)
gbm = GradientBoostingClassifier(n_estimators=100, learning_rate=0.1, random_state=42)

In [ ]:
def single_feature(c_model):
  for k in heart_data.columns.drop(target):
    X = heart_data[[k]]
    y = heart_data[target]
    #f1 score
    f1 = 0
    #splitting data
    aX_train, aX_test, ay_train, ay_test = train_test_split(X, y, test_size=0.2, random_state=42)
    #function for training and testing
    c_model.fit(aX_train, ay_train)
    ay_train_pred = c_model.predict(aX_train)
    ay_test_pred = c_model.predict(aX_test)
    f1 = f1_score(ay_test, ay_test_pred)
    #one model's results over different features
    print(k+":", f1)

In [ ]:
single_feature(rf)

age: 0.6381322957198444
sex: 0.7562500000000001
chest pain type: 0.7795275590551182
resting bp s: 0.4519230769230769
cholesterol: 0.6180257510729613
fasting blood sugar: 0.42222222222222217
resting ecg: 0.5407725321888412
max heart rate: 0.723076923076923
exercise angina: 0.7381974248927038
oldpeak: 0.7086614173228347
ST slope: 0.8074074074074076


In [ ]:
single_feature(gbm)

age: 0.640926640926641
sex: 0.7562500000000001
chest pain type: 0.7795275590551182
resting bp s: 0.6343283582089552
cholesterol: 0.688
fasting blood sugar: 0.42222222222222217
resting ecg: 0.5407725321888412
max heart rate: 0.7272727272727273
exercise angina: 0.7381974248927038
oldpeak: 0.7086614173228347
ST slope: 0.8074074074074076


In [ ]:
single_feature(et)

age: 0.6381322957198444
sex: 0.7562500000000001
chest pain type: 0.7795275590551182
resting bp s: 0.5405405405405406
cholesterol: 0.6422764227642276
fasting blood sugar: 0.42222222222222217
resting ecg: 0.5407725321888412
max heart rate: 0.6746031746031745
exercise angina: 0.7381974248927038
oldpeak: 0.7086614173228347
ST slope: 0.8074074074074076


##Concluding Evaluations

In [ ]:
from sklearn.metrics import classification_report

In [ ]:
heart_data.columns.drop(target)

Index(['age', 'sex', 'chest pain type', 'resting bp s', 'cholesterol',
       'fasting blood sugar', 'resting ecg', 'max heart rate',
       'exercise angina', 'oldpeak', 'ST slope'],
      dtype='object')

Lifestyle factor(s): exercise angina

###With lifestyle factor(s)

In [ ]:
#testing for the best mulitple features ranked by importance by that model, from a preivous section
all_columns = remov('exercise angina')
X = heart_data[all_columns]
y = heart_data[target]
#initializing 5 models
model = ExtraTreesClassifier(n_estimators=100, random_state=42, class_weight='balanced')
#splitting data
aX_train, aX_test, ay_train, ay_test = train_test_split(X, y, test_size=0.2, random_state=42)
#function for training and testing
model.fit(aX_train, ay_train)
ay_train_pred = model.predict(aX_train)
ay_test_pred = model.predict(aX_test)
with_f1 = f1_score(ay_test,ay_test_pred)

###Without lifestyle factor(s)

In [ ]:
#Using best model from preivous evaluations
all_columns = remov()
X = heart_data[all_columns]
y = heart_data[target]
#initializing 5 models
model = ExtraTreesClassifier(n_estimators=100, random_state=42, class_weight='balanced')
#splitting data
aX_train, aX_test, ay_train, ay_test = train_test_split(X, y, test_size=0.2, random_state=42)
#function for training and testing
model.fit(aX_train, ay_train)
ay_train_pred = model.predict(aX_train)
ay_test_pred = model.predict(aX_test)
without_f1 = f1_score(ay_test,ay_test_pred)

In [ ]:
print(f'{with_f1} (factors)\n{without_f1} (no factors)')

0.9425287356321839 (factors)
0.9420849420849421 (no factors)


###Findings (Lifestyle vs No Lifestyle Factors)
Removing exercise induced angina did not affect the result much, as it barely decreased the accuracy.

However, the feature importance of Extra Trees shows that it was the second most important feature, second to ST Slope.

Removing Other Most Important Features Ranked from Before for this model:

ST slope (1st), chest pain (3rd), old peak (4th)

In [ ]:
#Using best model from preivous evaluations
all_columns = remov('ST slope')
X = heart_data[all_columns]
y = heart_data[target]
#initializing 5 models
model = ExtraTreesClassifier(n_estimators=100, random_state=42, class_weight='balanced')
#splitting data
aX_train, aX_test, ay_train, ay_test = train_test_split(X, y, test_size=0.2, random_state=42)
#function for training and testing
model.fit(aX_train, ay_train)
ay_train_pred = model.predict(aX_train)
ay_test_pred = model.predict(aX_test)
print(f1_score(ay_test,ay_test_pred))

0.9140625000000001


In [ ]:
#Using best model from preivous evaluations
all_columns = remov('chest pain')
X = heart_data[all_columns]
y = heart_data[target]
#initializing 5 models
model = ExtraTreesClassifier(n_estimators=100, random_state=42, class_weight='balanced')
#splitting data
aX_train, aX_test, ay_train, ay_test = train_test_split(X, y, test_size=0.2, random_state=42)
#function for training and testing
model.fit(aX_train, ay_train)
ay_train_pred = model.predict(aX_train)
ay_test_pred = model.predict(aX_test)
print(f1_score(ay_test,ay_test_pred))

0.9420849420849421


In [ ]:
#Using best model from preivous evaluations
all_columns = remov('oldpeak')
X = heart_data[all_columns]
y = heart_data[target]
#initializing 5 models
model = ExtraTreesClassifier(n_estimators=100, random_state=42, class_weight='balanced')
#splitting data
aX_train, aX_test, ay_train, ay_test = train_test_split(X, y, test_size=0.2, random_state=42)
#function for training and testing
model.fit(aX_train, ay_train)
ay_train_pred = model.predict(aX_train)
ay_test_pred = model.predict(aX_test)
print(f1_score(ay_test,ay_test_pred))

0.9501915708812262


* Removing oldpeak improved the model performance by 1%
* Removing chest pain did not have much of an impact
* Removing ST slope decreased accuracy to 91%

Therefore, ST slope seems like an an important feature in predicting heart disease, as it was also ranked the most important feature of this model

Slope of the Peak Exercise ST Segment is a psycological measure that can indicate how well the heart is functioning under stress.